### DISCUSSION TITLE
Test resource cleanup on a per project basis?

### DISCUSSION THREAD

**Discussion by gzm0** [sbt/sbt#6603] [Q&A]
Context: https://github.com/scala-js/scala-js/issues/4317

Scala.js requires resource cleanup past test execution (it needs to terminate JS VMs).

We currently
- override `loadedTestFrameworks` and inject a JS specific bridge (https://github.com/scala-js/scala-js/blob/3b132da6fa12d0a048eaa9062159185ef20d4b4e/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala#L573-L611)
- use the global `onComplete` task to call their `close()` method to clean up (https://github.com/scala-js/scala-js/blob/3b132da6fa12d0a048eaa9062159185ef20d4b4e/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPlugin.scala#L351)

The huge upside of this is that we can fully re-use test detection and selection (e.g. `testQuick` works as you would expect).

However, this is problematic when running aggregate projects low parallelism: The tests run in sequence as expected, but the cleanup for the resources of an individual project's tests only happens after *all* the projects have completed testing. This can consume too much resources (e.g. too many open browser windows). The only workaround we currently have is to execute the projects individually.

Is there a better way to trigger resource cleanup in this scenario? Maybe something like `onComplete in <project>`?

**Comment by eed3si9n** [comment]
I don't remember the last time I used this thing, but the documentation does mention `Tests.Cleanup` - https://www.scala-sbt.org/1.x/docs/Testing.html#Setup+and+Cleanup

**Reply by gzm0** [reply]
This looked pretty much exactly what we needed. However, it turns out that this runs *before* [`done`](https://github.com/scala-js/scala-js/blob/3b132da6fa12d0a048eaa9062159185ef20d4b4e/test-interface/src/main/scala/sbt/testing/Runner.scala#L112) is called. So essentially it is too early for a cleanup hook of the test framework itself.
